### PR TITLE
Banking+App examples

### PR DESCRIPTION
**Pull request by skandtandon**
- Updated the Investment Banking example with relevant s3 minio changes
- Updated Airflow and Spark examples with minio changes

### CODE REVIEW

File: Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml [hunk 1]
```diff
@@ -8,14 +8,14 @@ spec:
...
-
-    # Needed if s3 server is behind proxy
-    # spark.hadoop.fs.s3a.proxy.host: "web-proxy.corp.hpecorp.net"
-    # spark.hadoop.fs.s3a.proxy.port: "8080"
-
-    # Needed if s3 server certs are not imported to java truststore used by spark
-    "spark.driver.extraJavaOptions": "-Dcom.amazonaws.sdk.disableCertChecking=true"
-    "spark.executor.extraJavaOptions": "-Dcom.amazonaws.sdk.disableCertChecking=true"
+    spark.hadoop.fs.s3a.access.key: <access-key>
+    spark.hadoop.fs.s3a.secret.key: <secret-key>
```
Comment (AlexNavara) [inline Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml:12]: For S3 proxy case this can be hardcoded to 's3'

File: Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml [hunk 2]
```diff
@@ -8,14 +8,14 @@ spec:
...
+    spark.hadoop.fs.s3a.s3.client.factory.impl: org.apache.hadoop.fs.s3a.NonChunkedDefaultS3ClientFactory
+    spark.driver.extraClassPath: /mounts/shared-volume/spark/disable-chunk-encoding-334.jar
+    spark.executor.extraClassPath: /mounts/shared-volume/spark/disable-chunk-encoding-334.jar
+    spark.hadoop.fs.s3a.bucket.ezua-demo.endpoint: http://local-s3-service.ezdata-system.svc.cluster.local:30000
+    spark.driver.extraJavaOptions: "-Dcom.amazonaws.sdk.disableCertChecking=true"
```
Comment (AlexNavara) [inline Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml:17]: 'disableCertChecking' options are not needed for s3-proxy case. There is no SSL

File: Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml [hunk 3]
```diff
@@ -8,14 +8,14 @@ spec:
     spark.hadoop.fs.s3a.path.style.access: "true"
     spark.hadoop.fs.s3a.impl: "org.apache.hadoop.fs.s3a.S3AFileSystem"
     spark.hadoop.fs.s3a.connection.ssl.enabled: "true"
```
Comment (AlexNavara) [inline Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml:10]: Disable ssl for s3-proxy

File: Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml [hunk 4]
```diff
@@ -24,33 +24,34 @@ spec:
...
   driver:
-    cores: 1
-    coreLimit: "1000m"
-    memory: "512m"
     labels:
       version: 3.3.1
     volumeMounts:
-      - mountPath: /mounts/data/
-        name: spark-data
+      - name: dev2-volume
```
Comment (AlexNavara) [inline Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml:39]: Here and further, what is 'dev2-volume'? I believe value with template '\<username\>-volume' should be used instead.

File: Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.4.1.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+apiVersion: "sparkoperator.hpe.com/v1beta2"
+kind: SparkApplication
+metadata:
+  name: data-process-transfer-demo
+  namespace: spark
```
Comment (AlexNavara) [inline Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.4.1.yaml:5]: Replace with '\<username\>' instead of 'spark'.

File: E2E-Demos/Investment-Banking/Airflow/spark-etl.py [hunk 6]
```diff
@@ -51,4 +56,4 @@ def end_job():
     python_callable=end_job,
     dag=dag,
 )
-task1>>task2>> task3
+task1>>task2>> task3
```
Comment (AlexNavara) [inline E2E-Demos/Investment-Banking/Airflow/spark-etl.py:59]: Don't forget about newlines in the end of files

File: E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml [hunk 7]
```diff
@@ -2,63 +2,45 @@ apiVersion: "sparkoperator.hpe.com/v1beta2"
...
-  namespace: "spark"
+  namespace: "hpedemo-user01"
 spec:
   sparkConf:
-    # S3 credentials should be specified below
-    spark.hadoop.fs.s3a.endpoint: "https://objectstore-zone1-svc.dataplatform.svc.cluster.local:9000/"
-    spark.hadoop.fs.s3a.access.key: "minioadmin"
-    spark.hadoop.fs.s3a.secret.key: "minioadmin"
-    spark.hadoop.fs.s3a.path.style.access: "true"
+    spark.executorEnv.SPARK_USER: hpedemo-user01
```
Comment (AlexNavara) [inline E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml:8]: Remove it. This is internal property set by autotix webhook

File: E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml [hunk 8]
```diff
@@ -2,63 +2,45 @@ apiVersion: "sparkoperator.hpe.com/v1beta2"
...
+    spark.hadoop.fs.s3a.access.key: <access-key>
+    spark.hadoop.fs.s3a.secret.key: <secret-key>
+    spark.hadoop.fs.s3a.endpoint: "ua01.hpe-minio-ezaf.com:9000"
```
Comment (AlexNavara) [inline E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml:11]: Shouldn't it be s3 proxy endpoint?

File: E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml [hunk 9]
```diff
@@ -2,63 +2,45 @@ apiVersion: "sparkoperator.hpe.com/v1beta2"
...
-      version: 3.3.1
-    cores: 1
-    coreLimit: "1"
-    memory: 1g
-    serviceAccount: hpe-spark
+      version: 3.4.1
     volumeMounts:
-      - name: {{dag_run.conf.get("username")}}-volume
-        mountPath: /mounts/{{dag_run.conf.get("username")}}-volume
+      - name: hpedemo-user01-volume
```
Comment (AlexNavara) [inline E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml:31]: I believe that Airflow must take care of template resolution, so use templates instead of hardcoding here and below.

File: E2E-Demos/Investment-Banking/Grafana/grafana.yaml [hunk 10]
```diff
@@ -7,7 +7,7 @@ spec:
         gateways:
         - istio-system/ezaf-gateway
         host:
-        - grafana.hpe-staging-ezaf.com
+        - grafana.hpe-apps-ezaf.com
```
Comment (AlexNavara) [inline E2E-Demos/Investment-Banking/Grafana/grafana.yaml:10]: Is this change needed?

**Review by AlexNavara — CHANGES_REQUESTED** [review]
See comments inline. There are a lot of duplicated mistakes, please make sure to review all examples for common mistakes and fix them all.

File: Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml (hunk 1 above)
Comment (akravacyber) [reply Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml:12]: Yes, access key should have JWT token, and secret key can have any value iianm, so leave 's3'

File: Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml [hunk 11]
```diff
@@ -2,39 +2,39 @@ apiVersion: "sparkoperator.hpe.com/v1beta2"
 kind: SparkApplication
 metadata:
   name: spark-fts-{{ts_nodash|replace("T", "")}}
+  namespace: spark
```
Comment (akravacyber) [inline Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml:5]: Namespace here is not needed, Airflow dynamically set user namespace during DAG execution.

```suggestion
```

File: Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.py [hunk 12]
```diff
@@ -22,7 +22,6 @@
     tags=['e2e example', 'ezaf', 'spark', 'csv', 'parquet', 'fts'],
     params={
         'training_path': Param("financial-processed", type="string"),
-        's3_secret_name': Param("spark-s3-creds", type="string"),
```
Comment (akravacyber) [inline Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.py:25]: Why did you remove it? Do you want to leave secret name hardcoded in the SparkApplication yaml file?

File: Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml [hunk 13]
```diff
@@ -2,39 +2,39 @@ apiVersion: "sparkoperator.hpe.com/v1beta2"
...
 spec:
   sparkConf:
-    spark.mapr.extraconf.secret: {{dag_run.conf.get("s3_secret_name", "spark-s3-creds")}}
     spark.hadoop.fs.s3a.path.style.access: "true"
     spark.hadoop.fs.s3a.impl: "org.apache.hadoop.fs.s3a.S3AFileSystem"
     spark.hadoop.fs.s3a.connection.ssl.enabled: "true"
+    spark.hadoop.fs.s3a.access.key: <access-token>
+    spark.hadoop.fs.s3a.secret.key: <secret-key>
```
Comment (akravacyber) [inline Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml:12]: It will not work. There are two ways: we add a new parameter for JWT token which we are passing in the params into this yaml. Or we continue to use the secret with s3 credentials.

File: E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml (hunk 8 above)
Comment (akravacyber) [reply E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml:11]: Yeah, agree. Here is the same comment. We need to use the secret with creds or pass s3 creds as DAG params and use here template

File: E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml (hunk 9 above)
Comment (akravacyber) [reply E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml:31]: Remove volumeMounts at all. Airflow is dynamically adding user, shared and datasources volumes for SparkApplication during DAG processing.

File: E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml [hunk 14]
```diff
@@ -2,63 +2,45 @@ apiVersion: "sparkoperator.hpe.com/v1beta2"
...
+        mountPath: /mnt/user
       - name: shared-volume
-        mountPath: /mounts/shared-volume
```
Comment (akravacyber) [inline E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml:56]: Same here, remove volume mounts

File: E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml [hunk 15]
```diff
@@ -2,63 +2,45 @@ apiVersion: "sparkoperator.hpe.com/v1beta2"
...
-    instances: 1
+        mountPath: /mnt/shared
   volumes:
-    - name: {{dag_run.conf.get("username")}}-volume
-      persistentVolumeClaim:
-        claimName: {{dag_run.conf.get("username")}}-spark-pvc
     - name: shared-volume
       persistentVolumeClaim:
-        claimName: kubeflow-shared-pvc
+        claimName: kubeflow-shared-pvc
```
Comment (akravacyber) [inline E2E-Demos/Investment-Banking/Airflow/spark_etl_new.yaml:46]: And please also remove volumes

**Review by akravacyber — CHANGES_REQUESTED** [review]

File: Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml [hunk 16]
```diff
@@ -4,37 +4,36 @@ metadata:
   name: spark-fts-{{ts_nodash|replace("T", "")}}
 spec:
   sparkConf:
-    spark.mapr.extraconf.secret: {{dag_run.conf.get("s3_secret_name", "spark-s3-creds")}}
     spark.hadoop.fs.s3a.path.style.access: "true"
     spark.hadoop.fs.s3a.impl: "org.apache.hadoop.fs.s3a.S3AFileSystem"
     spark.hadoop.fs.s3a.connection.ssl.enabled: "true"
+    spark.hadoop.fs.s3a.access.key: <access-token>
```
Comment (akravacyber) [inline Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml:10]: For SparkApplication yaml files in Airflow DAGs we couldn't leave some random placeholders, because Airflow doesn't know how to parse `<access-token>`. I guess you suppose that user will manually edit and paste here his JWT, but in Airflow user couldn't do it, because DAGs are pulled directly from git repo.

I propose to add a new param in python DAG file:
```python3
        'jwt_token': Param("", type="string"),
```

And use here a template for this param:
```suggestion
    spark.hadoop.fs.s3a.access.key: {{dag_run.conf.get("jwt_token", "")}}
```

**Review by akravacyber — CHANGES_REQUESTED** [review]

**Comment by skandtandon** [comment]
Updated the PR with the suggested changes.

File: Data-Analytics/Spark/devops/install.sh [hunk 17]
```diff
@@ -10,17 +10,17 @@ do
 done
 
 SCRIPTPATH=$(dirname "${0}")
-spark_version=${spark_version:=3.4.0}
-APP_IMAGE_NAME=${APP_IMAGE_NAME:-"gcr.io/mapr-252711/ezaf-spark-demo-example"}
+spark_version=${spark_version:=3.4.1}
+APP_IMAGE_NAME=${APP_IMAGE_NAME:-"gcr.io/mapr-252711/3.4.1"}
```
Comment (AlexNavara) [inline Data-Analytics/Spark/devops/install.sh:14]: I believe this is a typo. Image name should not be '3.4.1'

File: Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml [hunk 18]
```diff
@@ -24,33 +22,35 @@ spec:
...
-    cores: 1
-    coreLimit: "1000m"
-    memory: "512m"
     labels:
       version: 3.3.1
     volumeMounts:
-      - mountPath: /mounts/data/
-        name: spark-data
+      - name: <username>-volume
+        mountPath: /mounts/<username>-volume
```
Comment (AlexNavara) [inline Data-Analytics/Spark/k8s/DataProcessTransferFts-JarLocal-3.3.1.yaml:38]: You can leave the mount path static (do not include username there), as well as the volume name. Actually, only the PVC name matters.

File: Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml [hunk 19]
```diff
@@ -4,37 +4,36 @@ metadata:
   name: spark-fts-{{ts_nodash|replace("T", "")}}
 spec:
   sparkConf:
-    spark.mapr.extraconf.secret: {{dag_run.conf.get("s3_secret_name", "spark-s3-creds")}}
```
Comment (AlexNavara) [inline Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml:7]: Why did you decide to go with passing creds via sparkConf instead of using 'spark.mapr.extraconf.secret'? I believe previous comment from @akravacyber is also related to this.
If you use spark.mapr.extraconf.secret - you don't need to set access key and secret key explicitly in the yaml.

**Review by AlexNavara — CHANGES_REQUESTED** [review]
Main points:
1. Use 'spark.mapr.extraconf.secret' instead of passing s3 creds directly in the yaml file
2. Do not use 'username' placeholder in volume names and mountpoints. This just brings more unneeded complexity.

File: Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml (hunk 19 above)
Comment (akravacyber) [reply Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml:7]: Agree, I think it's better to use secret, rather than directly pass jwt token. Secret can be manually created from the kubeflow notebook

Comment (apurwaj2) [reply Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml:7]: Spark S3 secret notebook is already present. Please refer https://github.com/HPEEzmeral/ezua-tutorials/tree/release/fy23-q4/Data-Analytics/Spark-S3-secret

File: Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml [hunk 20]
```diff
@@ -4,37 +4,35 @@ metadata:
...
     spark.hadoop.fs.s3a.path.style.access: "true"
     spark.hadoop.fs.s3a.impl: "org.apache.hadoop.fs.s3a.S3AFileSystem"
     spark.hadoop.fs.s3a.connection.ssl.enabled: "true"
+    spark.mapr.extraconf.secret: "<k8s-secret-name>"
```
Comment (akravacyber) [inline Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.yaml:10]: Let's use template here and in `Data-Engineering/Airflow/example_ezaf_spark_csv_to_parquet_fts.py` we can return back `'s3_secret_name': Param("spark-s3-creds", type="string"),` and remove `jwt_token` parameter in such case. I think it's relevant to any Airflow DAG
```suggestion
    spark.mapr.extraconf.secret: {{dag_run.conf.get("s3_secret_name", "spark-s3-creds")}}  
```

**Review by akravacyber — COMMENTED** [review]
Overall looks good, some small nits

**Review by AlexNavara — APPROVED** [review]
lgtm

**Review by akravacyber — APPROVED** [review]
LGTM

**Review by prasadadireddi — APPROVED** [review]
lgtm
